### PR TITLE
Face Recognition Added

### LINKED ISSUE #456: Face Detection with OpenCV's built-in recognizer
Aim: This script recognizes the face in an image using OpenCV's built-in recognizer. 
Feature: Easy and understandable code 

I would like to work on this. Kindly assign me.

### PR DESCRIPTION
**Pull request by aakriti1318**
# Description

This script recognizes the face in an image using OpenCV's built-in recognizer.

## Have you read the [Contributing Guidelines on Pull Requests](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/CONTRIBUTING.md)?

- [Yes] Yes
- [ ] No

## Type of change

_Please delete options that are not relevant._

- [Yes] New feature (non-breaking change which adds functionality)
- [No] This change requires a documentation update

## Checklist:

- [ Yes] My code follows the style guidelines(Clean Code) of this project
- [ Yes] I have performed a self-review of my own code
- [ Yes] I have commented my code, particularly in hard-to-understand areas
- [ Yes] I have created a helpful and easy to understand `README.md`
- [Yes ] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)
- [Yes ] My changes generate no new warnings
- [Yes] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by aakriti1318** [comment]
Hey @avinashkranjan  could you please tell me what is this issue with my commits?

**Review by antrikshmisri — CHANGES_REQUESTED** [review]
Please add a `.gitignore` in your script folder and add all `.npy` , `.bin` files in it , also add a README.md.

**Comment by kaustubhgupta** [comment]
Hi @aakriti1318, 
To pass the Lint tests, Kindly do the following:
- autopep8 using pip
- Go to your project directory.
- Run the command: 
    ```bash 
       autopep8 -i -r --verbose ./
    ```

This should help you pass the test, after that I will review your code 👍🏼

**Comment by aakriti1318** [comment]
Hey @kaustubhgupta 
It is not working even after following all the steps. I am trying to commit 2 .py files, 2 .npy files, 2 .yml files, and 1 readme.md file. If there is any issue with any of the format then please tell me so that I can commit rest of the files.

**Comment by kaustubhgupta** [comment]
> Hey @kaustubhgupta
> It is not working even after following all the steps. I am trying to commit 2 .py files, 2 .npy files, 2 .yml files, and 1 readme.md file. If there is any issue with any of the format then please tell me so that I can commit rest of the files.

Okay, I will let you any new method. If you want you can try making new PR.

File: Face_Recognition/face_train.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+            img_arr = cv.imread(img_path)
+            gray = cv.cvtColor(img_arr, cv.COLOR_BGR2GRAY)
+
+            face_rect = haar_cascade.detectMultiScale(
+                gray, scaleFactor=1.1, minNeighbors=4)
+
+            for (x, y, w, h) in face_rect:
+                face_roi = gray[y:y+h, x:x+w]
+                features.append(face_roi)
```
Comment (antrikshmisri) [inline Face_Recognition/face_train.py:38]:
```suggestion
                features.append(face_roi)
```

**Comment by aakriti1318** [comment]
Okay sure, I have made a new repo <a href = "https://github.com/aakriti1318/Face-Recognition" >here </a>for the same PR I am trying. I would be grateful if you can review this.

**Comment by antrikshmisri** [comment]
@aakriti1318 The linter is failing because of the following lines:-
https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/77df39ea63a25bf30777e202e391ba643c4a44a4/Face_Recognition/face_train.py#L38-L39
Here is the error message:-
`Face_Recognition/face_train.py:39:16: E1101: Instance of 'ndarray' has no 'append' member (no-member)`
You might want to look into the logic

**Comment by aakriti1318** [comment]
> @aakriti1318 The linter is failing because of the following lines:-
> https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/77df39ea63a25bf30777e202e391ba643c4a44a4/Face_Recognition/face_train.py#L38-L39
> 
> 
> Here is the error message:-
> `Face_Recognition/face_train.py:39:16: E1101: Instance of 'ndarray' has no 'append' member (no-member)`
> You might want to look into the logic


This logic works fine in the local pc, but it fails here. This might be because the function dynamically adds the label corresponding to the face. 
I could comment this out, with a comment to remove the comment so that while running it won't give an error.

**Comment by antrikshmisri** [comment]
@aakriti1318 Can you cofirm if the directory structure should be like this:-
Screenshot (164)
If so I think there is a bug in your code

**Comment by aakriti1318** [comment]
> @aakriti1318 Can you cofirm if the directory structure should be like this:-
> Screenshot (164)
> If so I think there is a bug in your code

My directory look like <a href = "https://github.com/aakriti1318/Face-Recognition">this.</a>

File: Face_Recognition/face_train.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    people.append(i)
+
+haar_cascade = cv.CascadeClassifier('haar_face.xml')
+
+# For every face in the list the corresponding labels to the features are designated.
+features = []
+labels = []
+
+
+def train_face():
```
Comment (antrikshmisri) [inline Face_Recognition/face_train.py:22]:
```suggestion
def train_face(features:list):
```

File: Face_Recognition/face_train.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                continue
+            gray = cv.cvtColor(img_arr, cv.COLOR_BGR2GRAY)
+
+            face_rect = haar_cascade.detectMultiScale(gray, scaleFactor=1.1, minNeighbors=4)
+
+            for (x, y, w, h) in face_rect:
+                face = gray[y:y+h, x:x+w]
+                # remove below 2 lines comment
+                #features.append(face)
+                #labels.append(label)
```
Comment (antrikshmisri) [inline Face_Recognition/face_train.py:41]:
```suggestion
                # remove below 2 lines comment
                features.append(face)
                labels.append(label)
```

File: Face_Recognition/face_train.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+
+train_face()
```
Comment (antrikshmisri) [inline Face_Recognition/face_train.py:44]:
```suggestion
train_face(features)
```

File: Face_Recognition/Recognition.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+DIR = r'images'  # DIR - path of the folder containing images
+for i in os.listdir(DIR):
+    people.append(i)
+
+# LBPHFaceRecognizer - instatiate the face recognizer
+face_recognizer = cv.face.LBPHFaceRecognizer_create()
+# We are using the trained yml file.
+face_recognizer.read('face_trained.yml')
+
+# img = cv.imread(r'images\henry\13.jpg')
```
Comment (antrikshmisri) [inline Face_Recognition/Recognition.py:16]:
```suggestion
img = cv.imread(r'images\henry\13.jpg')
```

**Review by antrikshmisri — CHANGES_REQUESTED** [review]
@aakriti1318 Try the suggested changes and also delete all .npy/.bin files from the repo and add this line in .gitignore
`*.npy`

**Comment by kaustubhgupta** [comment]
@aakriti1318 try @antrikshmisri suggestions. If that doesn't work then create a new PR. In case new PR doesn't pass the case then we will merge it as it is because I have reviewed the code

File: Face_Recognition/Recognition.py (hunk 5 above)
Comment (aakriti1318) [reply Face_Recognition/Recognition.py:16]: I have tried this, but the same error is occurring.

**Comment by aakriti1318** [comment]
> @aakriti1318 try @antrikshmisri suggestions. If that doesn't work then create a new PR. In case new PR doesn't pass the case then we will merge it as it is because I have reviewed the code

Okay, I will try by creating a new PR and if that doesn't work then I will change the code also.

**Pull request closed without merging**